crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -304,49 +305,196 @@ fn build_new_replicated_loglet_configuration(
+        self.0.len()
+    }
+}
+
+/// Placement selector for picking a set of storage nodes for a replicated loglet out of a
+/// broader pool of available nodes.
+///
+/// This selector can be reused once constructed to make multiple placements in a single scheduling iteration
+/// if the node configuration and the replication settings are not changing.
+#[cfg(feature = "replicated-loglet")]
+#[derive(Clone)]
+struct LogletPlacementSelector<'a> {
+    nodes_config: &'a NodesConfiguration,
+    cluster_state: &'a ObservedClusterState,
+}
 
-    if log_servers.len() >= 3 {
-        let replication = ReplicationProperty::new(NonZeroU8::new(2).expect("to be valid"));
-        let mut nodeset = NodeSet::empty();
+impl<'a> LogletPlacementSelector<'a> {
+    fn new<'b: 'a>(
+        nodes_config: &'b NodesConfiguration,
+        cluster_state: &'b ObservedClusterState,
+    ) -> LogletPlacementSelector<'a> {
+        Self {
+            nodes_config,
+            cluster_state,
+        }
+    }
 
-        for node in &log_servers {
-            nodeset.insert(node.as_plain());
+    /// Picks a set of storage nodes for a replicated loglet out of the available pool. Only writable
+    /// nodes will be considered.
+    fn select<R: Rng + ?Sized>(
+        &self,
+        strategy: PlacementStrategy,
+        replication_property: &ReplicationProperty,
+        rng: &mut R,
+        preferred_nodes: &NodeSet,
+    ) -> Result<NodeSet, PlacementError> {
+        if replication_property.at_greatest_scope().0 != &LocationScope::Node {
+            // todo: add support for zonal and regional placement
+            unimplemented!("only node-scoped replication placement is currently supported");
         }
 
-        Some(ReplicatedLogletParams {
-            loglet_id,
-            sequencer: **log_servers
-                .iter()
-                .choose(&mut rng)
-                .expect("one node must exist"),
-            replication,
-            nodeset,
-            write_set: None,
-        })
-    } else if let Some(sequencer) = log_servers.iter().choose(&mut rng) {
-        previous_configuration.cloned().map(|mut configuration| {
-            configuration.loglet_id = loglet_id;
-            configuration.sequencer = **sequencer;
-            configuration
-        })
-    } else {
-        None
+        // Only consider alive, writable nodes.
+        let candidates = WritableNodeSet::from(self.cluster_state, self.nodes_config);
+
+        if candidates.len() < replication_property.num_copies().into() {
+            return Err(PlacementError::InsufficientWriteableNodes);
+        }
+
+        let placement = match strategy {
+            PlacementStrategy::AllAvailable => NodeSet::from_iter(candidates),
+            PlacementStrategy::CappedNodesetSize(size) => {
+                let mut nodes = preferred_nodes
+                    .iter()
+                    .copied()
+                    .choose_multiple(rng, size.get() as usize);
```

Comment:
Could it be that the preferred_nodes are currently not writable or don't have the `LogServer` role?

Reply:
Oh no! Good catch, this was definitely meant to be pre-filtered. I wrote `test_select_log_servers_respects_preferred_nodes` below but it's using the `CappedNodesetSize` policy and passing as a result. Will fix!

Reply:
Pff, noob mistake 😆 in the test for `CappedNodesetSize`, I had the check as:

```
        assert_that!(placement, contains(eq(PlainNodeId::new(3))));
        assert_that!(placement, contains(not(eq(PlainNodeId::new(1)))));
        assert_that!(placement, contains(not(eq(PlainNodeId::new(2)))));
```

should have been:

```
        assert_that!(placement, contains(eq(PlainNodeId::new(3))));
        assert_that!(placement, not(contains(eq(PlainNodeId::new(1)))));
        assert_that!(placement, not(contains(eq(PlainNodeId::new(2)))));
```